components.go

```diff
@@ -0,0 +1,149 @@
+	ButtonComponent     ComponentType = 2
+)
+
+// MessageComponent is a base interface for all message components.
+type MessageComponent interface {
+	json.Marshaler
+	Type() ComponentType
+}
+
+type unmarshalableMessageComponent struct {
+	MessageComponent
+}
+
+// UnmarshalJSON is a helper function to unmarshal MessageComponent object.
+func (umc *unmarshalableMessageComponent) UnmarshalJSON(src []byte) (err error) {
+	var v struct {
+		Type ComponentType `json:"type"`
+	}
+	err = json.Unmarshal(src, &v)
+	if err != nil {
+		return
+	}
+
+	var data MessageComponent
+	switch v.Type {
+	case ActionsRowComponent:
+		v := ActionsRow{}
+		err = json.Unmarshal(src, &v)
+		data = v
+	case ButtonComponent:
+		v := Button{}
+		err = json.Unmarshal(src, &v)
+		data = v
+	}
+	if err != nil {
+		return
+	}
+	umc.MessageComponent = data
+	return
+}
+
+// ActionsRow is a container for components within one row.
+type ActionsRow struct {
+	Components []MessageComponent `json:"components"`
+}
+
+// MarshalJSON is a method for marshaling ActionsRow to a JSON object.
+func (r ActionsRow) MarshalJSON() ([]byte, error) {
+	type actionsRow ActionsRow
+
+	return json.Marshal(struct {
+		actionsRow
+		Type ComponentType `json:"type"`
+	}{
+		actionsRow: actionsRow(r),
+		Type:       r.Type(),
+	})
+}
+
+// UnmarshalJSON is a helper function to unmarshal Actions Row.
+func (r *ActionsRow) UnmarshalJSON(data []byte) (err error) {
+	var v struct {
+		RawComponents []unmarshalableMessageComponent `json:"components"`
+	}
+	err = json.Unmarshal(data, &v)
+	if err != nil {
+		return
+	}
+	r.Components = make([]MessageComponent, len(v.RawComponents))
+	for i, v := range v.RawComponents {
+		r.Components[i] = v.MessageComponent
+	}
+	return
+}
+
+// Type is a method to get the type of a component.
+func (r ActionsRow) Type() ComponentType {
+	return ActionsRowComponent
+}
+
+// ButtonStyle is style of button.
+type ButtonStyle uint
+
+// Button styles.
+const (
+	// PrimaryButton is a button with blurple color.
+	PrimaryButton ButtonStyle = 1
+	// SecondaryButton is a button with grey color.
+	SecondaryButton ButtonStyle = 2
+	// SuccessButton is a button with green color.
+	SuccessButton ButtonStyle = 3
+	// DangerButton is a button with red color.
+	DangerButton ButtonStyle = 4
+	// LinkButton is a special type of button which navigates to a URL. Has grey color.
+	LinkButton ButtonStyle = 5
+)
+
+// ButtonEmoji represents button emoji, if it does have one.
+type ButtonEmoji struct {
+	Name     string `json:"name,omitempty"`
+	ID       string `json:"id,omitempty"`
+	Animated bool   `json:"animated,omitempty"`
+}
+
+// Button represents button component.
+type Button struct {
+	Label    string      `json:"label"`
+	Style    ButtonStyle `json:"style"`
+	Disabled bool        `json:"disabled"`
+	Emoji    ButtonEmoji `json:"emoji"`
+
+	// NOTE: Only button with LinkButton style can have link. Also, Link is mutually exclusive with CustomID.
+	Link     string `json:"url,omitempty"`
```

Comment:
A bit of a nit, but is there any particular reason this field name deviates from `URL`? I don't see a compelling reason to go against what will be in the Discord documentation (as I discovered in testing this!).